### PR TITLE
[FLINK-3806] [gelly] Revert use of DataSet.count()

### PR DESCRIPTION
**Pull request by greghogan**
This leaves the Graph API unchanged but GatherSumApplyIteration and ScatterGatherIteration now use broadcast variables to share the numberOfVertices count. The PageRanks have been updated to use this feature.

### CODE REVIEW

File: flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/gsa/GatherSumApplyIteration.java [hunk 1]
```diff
@@ -289,6 +300,11 @@ private GatherUdf(GatherFunction<VV, EV, M> gatherFunction, TypeInformation<Tupl
 
 		@Override
 		public void open(Configuration parameters) throws Exception {
+			try {
+				Collection<LongValue> numberOfVertices = getRuntimeContext().getBroadcastVariable("number of vertices");
+				this.gatherFunction.setNumberOfVertices(numberOfVertices.iterator().next().getValue());
+			} catch (Exception e) {
+			}
```
Comment (vasia) [inline flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/gsa/GatherSumApplyIteration.java]: I don't think we should leave the catch block empty here. Can we provide a meaningful message instead?

File: flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/gsa/GatherSumApplyIteration.java [hunk 2]
```diff
@@ -327,6 +343,11 @@ private SumUdf(SumFunction<VV, EV, M> sumFunction, TypeInformation<Tuple2<K, M>>
 
 		@Override
 		public void open(Configuration parameters) throws Exception {
+			try {
+				Collection<LongValue> numberOfVertices = getRuntimeContext().getBroadcastVariable("number of vertices");
+				this.sumFunction.setNumberOfVertices(numberOfVertices.iterator().next().getValue());
+			} catch (Exception e) {
+			}
```
Comment (vasia) [inline flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/gsa/GatherSumApplyIteration.java]: Likewise here and below.

**Comment by vasia** [comment]
Thanks for the PR @greghogan! I left one comment.
The HITS implementation also makes use of the number of vertices in the same way PageRank did. We can update that library method too. Finally, could you update the usage docs of PageRank to remove `numVertices` from the parameters?
Thank you!

File: flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/gsa/GatherSumApplyIteration.java (hunk 1 above)
Comment (greghogan) [inline flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/gsa/GatherSumApplyIteration.java]: Could also add a hasBroadcastVariable method to RuntimeContext. I had first expected that getBroadcastVariable would return an empty Collection.

Comment (StephanEwen) [inline flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/gsa/GatherSumApplyIteration.java]: The BatchAPI usually throws exceptions when people build a seemingly incorrect program, for example by using a broadcast variable that does not exist. It seemed to help people to catch bugs faster.

The exception makes it obvious that something is wrong, the empty set would make various programs run but return a wrong result. That is much harder to recognize and debug.

Comment (StephanEwen) [inline flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/gsa/GatherSumApplyIteration.java]: BTW: In the IntelliJ code style, they encourage to call ignored exceptions "ignored", or to at least add a comment why the exception is ignored:
```java
try {
    ...
} catch (Exception ignored) {}
```
or
```java
try {
    ...
} catch (Exception e) {
    // thrown if BV does not exist, ignore here
```

Comment (greghogan) [inline flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/gsa/GatherSumApplyIteration.java]: @sewen thanks for the pointers.

What are your thoughts on adding a hasBroadcastVariable method to RuntimeContext? And elsewhere in that class?

The javadocs for getBroadcastVariable could note the exception as for getAccumulator.

Comment (StephanEwen) [inline flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/gsa/GatherSumApplyIteration.java]: I think that makes sense.

**Comment by greghogan** [comment]
All comments should now be addressed.

**Comment by vasia** [comment]
Thanks @greghogan! It looks like you have some checkstyle violation, otherwise +1.

**Comment by greghogan** [comment]
Yes. If I allow IntelliJ to handle automatically organize imports then they can be reordered due to a lack of consistency and project-wide guidelines.

**Comment by greghogan** [comment]
Travis CI is now green.

**Comment by greghogan** [comment]
Will merge this ...

**Pull request closed without merging**